src/main/java/org/opentripplanner/routing/stoptimes/AlternativeLegs.java

```diff
@@ -0,0 +1,242 @@
+import org.opentripplanner.model.TripTimeOnDate;
+import org.opentripplanner.model.calendar.ServiceDate;
+import org.opentripplanner.model.plan.Leg;
+import org.opentripplanner.model.plan.ScheduledTransitLeg;
+import org.opentripplanner.routing.RoutingService;
+import org.opentripplanner.routing.core.ServiceDay;
+import org.opentripplanner.routing.trippattern.TripTimes;
+
+/**
+ * A helper class to fetch previous/next alternative legs for a scheduled transit leg.
+ * The replacement legs arrive/depart from/to the same station as the original leg, but uses other
+ * trips for the leg.
+ *
+ * Generalized cost and constrained transfers are not included in the alternative legs.
+ */
+public class AlternativeLegs {
+
+  public static final int ZERO_COST = 0;
+
+  public static List<ScheduledTransitLeg> getAlternativeLegs(
+    Leg leg,
+    Integer numberLegs,
+    RoutingService routingService,
+    boolean searchBackward
+  ) {
+    StopLocation fromStop = leg.getFrom().stop;
+    StopLocation toStop = leg.getTo().stop;
+
+    Station fromStation = fromStop.getParentStation();
+    Station toStation = toStop.getParentStation();
+
+    Collection<StopLocation> origins = fromStation == null
+      ? List.of(fromStop)
+      : fromStation.getChildStops();
+
+    Collection<StopLocation> destinations = toStation == null
+      ? List.of(toStop)
+      : toStation.getChildStops();
+
+    TimetableSnapshot timetableSnapshot = routingService.getTimetableSnapshot();
+
+    Comparator<ScheduledTransitLeg> legComparator = Comparator.comparing(
+      ScheduledTransitLeg::getStartTime
+    );
+
+    if (searchBackward) {
+      legComparator = legComparator.reversed();
+    }
+
+    return origins
+      .stream()
+      .flatMap(stop -> routingService.getPatternsForStop(stop, timetableSnapshot).stream())
+      .filter(tripPattern -> tripPattern.getStops().stream().anyMatch(destinations::contains))
+      .flatMap(tripPattern -> withBoardingAlightingPositions(origins, destinations, tripPattern))
+      .flatMap(t ->
+        generateLegs(
+          routingService,
+          timetableSnapshot,
+          t,
+          leg.getStartTime(),
+          leg.getServiceDate(),
+          searchBackward
+        )
+      )
+      .filter(Predicate.not(leg::isPartiallySameTransitLeg))
+      .sorted(legComparator)
+      .limit(numberLegs)
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * This has been copied and slightly modified from StopTimesHelper.
+   * TODO: Adapt after new transit model is in place
+   */
+  @Nonnull
+  private static Stream<ScheduledTransitLeg> generateLegs(
+    RoutingService routingService,
+    TimetableSnapshot timetableSnapshot,
+    T2<TripPattern, P2<Integer>> patternWithBoardAlightPositions,
+    Calendar departureTime,
+    ServiceDate originalDate,
+    boolean searchBackward
+  ) {
+    TripPattern pattern = patternWithBoardAlightPositions.first;
+    int boardingPosition = patternWithBoardAlightPositions.second.first;
```

Comment:
I know this might change in the future but this `.second.first` is a bit ambiguous

Reply:
I added custom record types for these.